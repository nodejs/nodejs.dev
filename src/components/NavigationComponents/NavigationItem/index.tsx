import React from 'react';
import { LocalizedLink as Link } from 'gatsby-theme-i18n';
import classnames from 'classnames';
import { useScrollToElement } from '../../../hooks/useScrollToElement';
import { useMediaQuery } from '../../../hooks/useMediaQuery';
import styles from './index.module.scss';

interface Props {
  slug: string;
  title: string;
  isActive: boolean;
  extraClasses?: string;
  onClick?: (event: React.MouseEvent<HTMLLinkElement>) => void;
}

const NavigationItem = ({
  slug,
  title,
  onClick,
  isActive,
  extraClasses,
}: Props): JSX.Element => {
  const className = classnames('t-body2', styles.navigationItem, extraClasses, {
    [styles.navigationItemActive]: isActive,
  });

  const isMobile = useMediaQuery('(max-width: 900px)');

  useScrollToElement({
    element: slug,
    smooth: true,
    containerId: 'main-navigation',
    offset: -100,
    // This needs to be a strict `===false` as
    // the initial value is `undefined` as being CSR
    isActive: isActive && isMobile === false,
  });

  if (slug.startsWith('https')) {
    return (
      <a id={slug} href={slug} className={className}>
        {title}
      </a>
    );
  }

  return (
    <Link
      id={slug}
      to={slug}
      className={className}
      onClick={onClick}
      aria-current={isActive ? 'page' : undefined}
    >
      {title}
    </Link>
  );
};

export default NavigationItem;
