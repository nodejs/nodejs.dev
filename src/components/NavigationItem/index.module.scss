.navigation-item,
%navigation-item {
  align-items: center;
  border-radius: 4px;
  box-sizing: border-box;
  color: var(--color-text-secondary);
  cursor: pointer;
  display: block;
  list-style: none;
  margin: 0 0 0 var(--space-24);
  padding: calc(var(--space-12) / 2) 0 calc(var(--space-24) / 2) var(--space-08);
  position: relative;
  text-decoration: none;

  &:last-of-type::after {
    display: none;
  }

  &::before {
    background-size: 24px;
    border-radius: 50%;
    box-shadow: inset 0 0 0 3px var(--color-border-primary);
    box-sizing: border-box;
    content: ' ';
    display: block;
    height: 24px;
    left: -24px;
    overflow: hidden;
    padding: 5px 4px;
    position: absolute;
    top: 0.6rem;
    transition: box-shadow 0.28s;
    width: 24px;
  }

  &--done::before {
    content: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' fill='white' viewBox='0 0 24 24'%3E%3Cpath d='M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z'%3E%3C/path%3E%3C/svg%3E");
  }

  &--active {
    color: var(--color-text-accent);
    font-weight: 600;
  }

  &--done::before,
  &--active::before {
    box-shadow: inset 0 0 0 2.4rem var(--color-border-accent);
  }

  &:hover::before {
    box-shadow: inset 0 0 0 2.4rem var(--color-border-accent);
  }

  &::after {
    background: var(--color-border-primary);
    content: ' ';
    display: block;
    height: calc(100% - 1.2rem - 12px);
    left: calc(-1.2rem - 1px);
    position: absolute;
    top: calc(2.4rem + 6px);
    width: 2px;
  }

  &--done::after {
    background: var(--color-border-accent);
  }

  &--api {
    padding: var(--space-08) 0 0 0;

    &:hover::before,
    &::before {
      box-shadow: none;
    }

    &:after {
      content: none;
    }
  }
}
