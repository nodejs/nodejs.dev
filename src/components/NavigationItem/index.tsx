import React from 'react';
import { Link } from 'gatsby';
import classnames from 'classnames';

interface Props {
  key: string;
  isLearn?: boolean;
  isRead: boolean;
  isActive: boolean;
  slug: string;
  title: string;
  onClick?: () => void;
}

const NavigationItem = ({
  isLearn = true,
  isRead,
  isActive,
  slug,
  title,
  onClick,
}: Props): JSX.Element => {
  const className = isLearn
    ? classnames('t-body2 side-nav__item', {
        'side-nav__item--done': isRead,
        'side-nav__item--active': !isRead && isActive,
      })
    : classnames('t-body2 side-nav__item-community', {
        'side-nav__item-community--active': isActive,
      });

  return (
    <Link
      to={`/learn/${slug}`}
      id={`link-${slug}`}
      onClick={onClick}
      className={className}
    >
      {title}
    </Link>
  );
};

export default NavigationItem;
