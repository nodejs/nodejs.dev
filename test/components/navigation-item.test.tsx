import React from 'react';
import renderer from 'react-test-renderer';
import NavigationItem from '../../src/components/navigation-item';

describe('NavigationItem component', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <NavigationItem
          key={'123'}
          isRead={false}
          isActive={true}
          slug={'versioning'}
          title={'Versioning'}
          onClick={console.log}
          autoScroll={console.log}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
