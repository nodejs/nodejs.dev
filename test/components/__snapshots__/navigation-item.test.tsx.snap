// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationItem component renders correctly 1`] = `
<a
  className="side-nav__item side-nav__item--active"
  href="/versioning"
  onClick={[Function]}
>
  Versioning
</a>
`;
